server/src/main/java/com/cloud/api/query/QueryManagerImpl.java

```diff
@@ -4690,11 +4645,28 @@ private Pair<List<SnapshotJoinVO>, Integer> searchForSnapshotsWithParams(final L
             sb.join("tagSearch", tagSearch, sb.entity().getId(), tagSearch.entity().getResourceId(), JoinBuilder.JoinType.INNER);
         }
 
+        if (storagePoolId != null) {
+            SearchBuilder<SnapshotDataStoreVO> storagePoolSb = snapshotDataStoreDao.createSearchBuilder();
+            storagePoolSb.and("poolId", storagePoolSb.entity().getDataStoreId(), SearchCriteria.Op.EQ);
+            storagePoolSb.and("role", storagePoolSb.entity().getRole(), SearchCriteria.Op.EQ);
+            sb.join("storagePoolSb", storagePoolSb, sb.entity().getId(), storagePoolSb.entity().getSnapshotId(), JoinBuilder.JoinType.INNER);
+        }
```

Comment:
@vishesh92 can't we do something like the following?
        if (storagePoolId != null) {
            sc.setParameters("imageStoreId", storagePoolId); // change key for imageStoreId to storeId
            sc.setParameters("locationType", "Primary");
        }

Reply:
It won't behave as expected, if the user passes in both imagestore and storagepool.
That is why for `imagestoreId`, I am querying within the view and joining with storagepool for `storagepoolId`.

Reply:
complicated. :brain: 
It is what it is